File: packages/edge/src/services/tokenUsageTracker.ts
import axios from 'axios';
import { config } from '../config/env';
import logger from '../config/logger';

interface TokenUsageStats {
  usageCount: number;
  lastUsedAt: Date;
}

/**
 * TokenUsageTracker - Tracks API token usage in Edge server
 * Periodically reports usage statistics to backend for aggregation
 */
class TokenUsageTracker {
  private usageMap: Map<number, TokenUsageStats> = new Map(); // tokenId (number) -> stats
  private reportIntervalMs: number;
  private reportTimer: NodeJS.Timeout | null = null;
  private initialized = false;
  private readonly edgeInstanceId: string;

  constructor() {
    this.reportIntervalMs = parseInt(process.env.TOKEN_USAGE_REPORT_INTERVAL_MS || '60000');
    this.edgeInstanceId = `edge-${config.group}-${process.pid}`;
  }

  /**
   * Initialize the usage tracker
   * Starts periodic reporting to backend
   */
  async initialize(): Promise<void> {
    if (this.initialized) {
      logger.warn('[TokenUsageTracker] Already initialized');
      return;
    }

    logger.info('[TokenUsageTracker] Initializing token usage tracker...', {
      reportIntervalMs: this.reportIntervalMs,
      edgeInstanceId: this.edgeInstanceId,
    });

    // Start periodic reporting
    this.reportTimer = setInterval(() => {
      this.reportUsageToBackend().catch(err => {
        logger.error('[TokenUsageTracker] Failed to report usage:', err.message);
      });
    }, this.reportIntervalMs);

    this.initialized = true;
    logger.info('[TokenUsageTracker] Initialized');
  }

  /**
   * Shutdown the usage tracker
   * Reports any remaining usage before shutdown
   */
  async shutdown(): Promise<void> {
    if (this.reportTimer) {
      clearInterval(this.reportTimer);
      this.reportTimer = null;
    }

    // Report remaining usage before shutdown
    if (this.usageMap.size > 0) {
      try {
        await this.reportUsageToBackend();
      } catch (error) {
        logger.error('[TokenUsageTracker] Failed to report usage during shutdown:', error);
      }
    }

    this.usageMap.clear();
    this.initialized = false;
    logger.info('[TokenUsageTracker] Shutdown complete');
  }

  /**
   * Record token usage
   * Called when a token is successfully validated and used
   * @param tokenId - The numeric token ID from the database
   */
  recordUsage(tokenId: number): void {
    const existing = this.usageMap.get(tokenId);
    const now = new Date();

    if (existing) {
      existing.usageCount++;
      existing.lastUsedAt = now;
    } else {
      this.usageMap.set(tokenId, {
        usageCount: 1,
        lastUsedAt: now,
      });
    }
  }

  /**
   * Report usage statistics to backend
   * Clears local usage data after successful report
   */
  private async reportUsageToBackend(): Promise<void> {
    if (this.usageMap.size === 0) {
      return;
    }

    // Collect and clear usage data atomically
    const usageData: Array<{
      tokenId: number;
      usageCount: number;
      lastUsedAt: string;
    }> = [];

    for (const [tokenId, stats] of this.usageMap) {
      usageData.push({
        tokenId,
        usageCount: stats.usageCount,
        lastUsedAt: stats.lastUsedAt.toISOString(),
      });
    }

    // Clear the map before sending to avoid losing data if request fails
    const backupMap = new Map(this.usageMap);
    this.usageMap.clear();

    try {
      const response = await axios.post(
        `${config.gatrixUrl}/api/v1/server/internal/token-usage-report`,
        {
          edgeInstanceId: this.edgeInstanceId,
          usageData,
          reportedAt: new Date().toISOString(),
        },
        {
          headers: {
            'x-api-token': config.apiToken,
            'x-application-name': config.applicationName,
          },
          timeout: 10000,
        }
      );

      if (response.data?.success) {
        logger.info('[TokenUsageTracker] Usage reported successfully', {
          tokenCount: usageData.length,
          totalUsage: usageData.reduce((sum, d) => sum + d.usageCount, 0),
        });
      } else {
        throw new Error(response.data?.message || 'Unknown error');
      }
    } catch (error: any) {
      // Restore usage data on failure
      for (const [tokenId, stats] of backupMap) {
        const existing = this.usageMap.get(tokenId);
        if (existing) {
          existing.usageCount += stats.usageCount;
          if (stats.lastUsedAt > existing.lastUsedAt) {
            existing.lastUsedAt = stats.lastUsedAt;
          }
        } else {
          this.usageMap.set(tokenId, stats);
        }
      }
      throw error;
    }
  }
}

export const tokenUsageTracker = new TokenUsageTracker();

File: packages/edge-dotnet/src/Gatrix.Edge/Services/TokenUsageTracker.cs
using System.Collections.Concurrent;
using System.Net.Http.Json;
using System.Text.Json;
using Gatrix.Edge.Options;
using Microsoft.Extensions.Options;

namespace Gatrix.Edge.Services;

/// <summary>
/// Tracks API token usage and periodically reports to backend.
/// </summary>
public class TokenUsageTracker : IHostedService, IDisposable
{
    private readonly ConcurrentDictionary<int, TokenUsageStats> _usageMap = new();
    private readonly EdgeOptions _options;
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly ILogger<TokenUsageTracker> _logger;
    private readonly string _edgeInstanceId;
    private Timer? _reportTimer;

    public TokenUsageTracker(
        IOptions<EdgeOptions> options,
        IHttpClientFactory httpClientFactory,
        ILogger<TokenUsageTracker> logger)
    {
        _options = options.Value;
        _httpClientFactory = httpClientFactory;
        _logger = logger;
        _edgeInstanceId = $"edge-{_options.Group}-{System.Diagnostics.Process.GetCurrentProcess().Id}";
    }

    public Task StartAsync(CancellationToken cancellationToken)
    {
        _logger.LogInformation("[TokenUsageTracker] Initializing (interval={Interval}ms, instance={Instance})",
            _options.TokenUsageReportIntervalMs, _edgeInstanceId);

        _reportTimer = new Timer(
            async _ => await ReportUsageToBackendSafe(),
            state: null,
            dueTime: _options.TokenUsageReportIntervalMs,
            period: _options.TokenUsageReportIntervalMs);

        _logger.LogInformation("[TokenUsageTracker] Initialized");
        return Task.CompletedTask;
    }

    public async Task StopAsync(CancellationToken cancellationToken)
    {
        _reportTimer?.Change(Timeout.Infinite, 0);

        if (!_usageMap.IsEmpty)
        {
            try { await ReportUsageToBackendAsync(); }
            catch (Exception ex) { _logger.LogError(ex, "[TokenUsageTracker] Failed to report during shutdown"); }
        }

        _usageMap.Clear();
        _logger.LogInformation("[TokenUsageTracker] Shutdown complete");
    }

    /// <summary>
    /// Record a token usage.
    /// </summary>
    public void RecordUsage(int tokenId)
    {
        _usageMap.AddOrUpdate(tokenId,
            _ => new TokenUsageStats { UsageCount = 1, LastUsedAt = DateTime.UtcNow },
            (_, existing) =>
            {
                existing.UsageCount++;
                existing.LastUsedAt = DateTime.UtcNow;
                return existing;
            });
    }

    private async Task ReportUsageToBackendSafe()
    {
        try { await ReportUsageToBackendAsync(); }
        catch (Exception ex) { _logger.LogError(ex, "[TokenUsageTracker] Failed to report usage"); }
    }

    private async Task ReportUsageToBackendAsync()
    {
        if (_usageMap.IsEmpty) return;

        // Snapshot and clear atomically
        var usageData = new List<object>();
        var backup = new Dictionary<int, TokenUsageStats>();

        foreach (var kvp in _usageMap)
        {
            if (_usageMap.TryRemove(kvp.Key, out var stats))
            {
                backup[kvp.Key] = stats;
                usageData.Add(new
                {
                    tokenId = kvp.Key,
                    usageCount = stats.UsageCount,
                    lastUsedAt = stats.LastUsedAt.ToString("o"),
                });
            }
        }

        try
        {
            var client = _httpClientFactory.CreateClient("GatrixBackend");
            var request = new HttpRequestMessage(HttpMethod.Post, "/api/v1/server/internal/token-usage-report")
            {
                Content = JsonContent.Create(new
                {
                    edgeInstanceId = _edgeInstanceId,
                    usageData,
                    reportedAt = DateTime.UtcNow.ToString("o"),
                })
            };
            request.Headers.Add("x-api-token", _options.ApiToken);
            request.Headers.Add("x-application-name", _options.ApplicationName);

            var response = await client.SendAsync(request);
            response.EnsureSuccessStatusCode();

            _logger.LogInformation("[TokenUsageTracker] Usage reported: {Count} tokens, {Total} total",
                usageData.Count, backup.Values.Sum(s => s.UsageCount));
        }
        catch
        {
            // Restore on failure
            foreach (var kvp in backup)
            {
                _usageMap.AddOrUpdate(kvp.Key,
                    _ => kvp.Value,
                    (_, existing) =>
                    {
                        existing.UsageCount += kvp.Value.UsageCount;
                        if (kvp.Value.LastUsedAt > existing.LastUsedAt)
                            existing.LastUsedAt = kvp.Value.LastUsedAt;
                        return existing;
                    });
            }
            throw;
        }
    }

    public void Dispose()
    {
        _reportTimer?.Dispose();
        GC.SuppressFinalize(this);
    }

    private class TokenUsageStats
    {
        public int UsageCount { get; set; }
        public DateTime LastUsedAt { get; set; }
    }
}
